app/pathfinder_svc.py

```diff
@@ -73,49 +74,116 @@ def add_fact(fact_list, trait, value):
         source.access = BaseWorld.Access.RED
         await self.data_svc.store(source)
         return source
+    
+    async def generate_path_analysis_report(self, vuln_report, start, target, whitelist: list, blacklist: list):  
+        """
+        """
+        r = dict()
+        r['exploitability_graph'] = await self.generate_exploitability_graph(vuln_report, whitelist, blacklist)
+        r['exploitability_paths'] = await self.generate_exploitable_paths(vuln_report, r['exploitability_graph'], start, target)
+        return r
 
-    async def generate_adversary(self, report, initial_host, target_host, tags=None):
-        async def create_cve_adversary(techniques, tags):
-            adv_id = uuid.uuid4()
-            obj_default = (
-                await self.data_svc.locate('objectives', match=dict(name='default'))
-            )[0]
-            return dict(
-                id=str(adv_id),
-                name='pathfinder adversary',
-                description='auto generated adversary for pathfinder',
-                atomic_ordering=techniques,
-                tags=tags,
-                objective=obj_default.id,
-            )
+    async def generate_exploitability_graph(self, report, whitelist, blacklist):
+        """
+            Explores and evaluates the nodes and edges of the report.network map. Removes any nodes that are on the blacklist or 
+            do not have any abilities that provide lateral movement.
+
+        Args:
+            report (): Caldera vulnerability report
+            whitelist (): list of nodes that are known owned or are considered freebies
+            blacklist (): list of nodes that should not be used during exploitability analysis
+        Returns
+            (NetworkX Graph)
 
-        def get_all_tags(objlist):
-            return [t for a in objlist for t in a.tags]
+        """
+        async def _is_edge_exploitable(candidate, whitelist, blacklist):
+            """
+                Determines if candidate node is exploitable. Exploitability on the
+                target node is defined as obtaining presence (either normal or privileged access) on the
+                candidate node.
+            Args:
+                candidate (): target node to determine exploitability to
+            Returns: 
+                (bool)
+            """
+            if not candidate.can_access():
+                return False
+            if candidate.is_denied():
+                return False
+            if candidate in blacklist:
+                return False
+            if candidate.freebie_abilities:
+                return True
+            if candidate.cves and await self.get_host_exploits(report, candidate):
+                return True
+            if candidate in whitelist:
+                return True
+            # Defaults to true since we want to mark nodes that aren't offlimits as a freebie potentially (in generate_exp_paths)
```

Comment:
This is a general comment, but line-length standardization would be nice to have before merge.

Reply:
I agree. I figured MVP would be working, and then I'd later go back through for cleanliness since code is pretty dynamic right now.